[git-webkit] Add trace command

JonWBedard opened the pull request:
#### d3e19e632f184c36329139ac86089f604afc477d
<pre>
[git-webkit] Add trace command
<a href="https://bugs.webkit.org/show_bug.cgi?id=244859">https://bugs.webkit.org/show_bug.cgi?id=244859</a>
&lt;rdar://97455957&gt;

Reviewed by Dewei Zhu.

* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/__init__.py: Add Trace command.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
(Pickable.pickable): Create a CommitsStory object to track more commit relationships.
(Pickable.main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/trace.py: Added.
(Relationship.reversed): Reverse the relationship type (ie, the reverse of &quot;reverted&quot; is &quot;reverted by&quot;).
(Relationship.parse): Parse a commit message and return relationships.
(Relationship.__init__):
(Relationship.__repr__):
(CommitsStory.issues_for): Parse a commit message for embedded issue references.
(CommitsStory.__init__):
(CommitsStory.add): Add a commit to the story.
(Trace.parser):
(Trace.relationships): Find all relationships for the specified commit.
(Trace.summary): Sumarize a commit.
(Trace.main): Find all relationships for the specified commit and print them.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/trace_unittest.py: Added.
(TestRelationship):
(TestCommitsStory):
(TestTrace):

Canonical link: <a href="https://commits.webkit.org/254997@main">https://commits.webkit.org/254997@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2837aed3c668d320b8b54a09bbd142e02745e543

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90969 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35540 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/21506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/100312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/158813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94975 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/34041 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29059 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83313 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/97105 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96625 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WK2-Tests-EWS "Waiting to run tests") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/77756 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/26928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/94548 "webkitpy-tests (failure)") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/81874 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/21506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/69963 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35121 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/21506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/90035 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32921 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/21506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36699 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/77756 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1528 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38626 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/21506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

alancoon commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
```diff
@@ -51,17 +52,36 @@ def parser(cls, parser, loggers=None):
     @classmethod
-    def pickable(cls, commits):
-        filtered_commits = []
+    def pickable(cls, commits, repository, commits_bundle=None):
```
What is a `commits_bundle`?

JonWBedard replied:
I'm open to other names for this object.

Basically, it is often the case that another commit has the information that defines a relationship for the commit in question. Take, for example, a commit which is reverted by a subsequent commit. The subsequent commit defines the relationship. The `commits_bundle` object basically has a bunch of pre-computed commit relationships in it so we can just look them up by key instead of searching through all commits. It basically turns `pickable` from `O(n^2)` to `O(n)`.

alancoon replied:
I think I'd suggest a name like `CommitsStory` rather than `CommitsBundle`, "bundle" didn't imply that the contained commits have a meaningful relationship.

dewei-zhu commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/trace.py:
```diff
@@ -0,0 +1,276 @@
+            cls.REVERTED_BY: cls.REVERTS,
+            cls.FOLLOW_UP: cls.FOLLOW_UP_BY,
+            cls.FOLLOW_UP_BY: cls.FOLLOW_UP,
+        }.get(type, type)
```
Is it intentional to not raise an error when type is unknown type?

alancoon commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
```diff
@@ -51,17 +52,36 @@ def parser(cls, parser, loggers=None):
-    def pickable(cls, commits):
-        filtered_commits = []
+    def pickable(cls, commits, repository, commits_story=None):
+        filtered_in = set()
```
I'd suggest a more descriptive name for `filtered_in` here, something that indicates that these are commits that were originally authored on the branch, such as `original_authored_commit_shas`.

alancoon commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
```diff
@@ -51,17 +52,36 @@ def parser(cls, parser, loggers=None):
+            for rel in relationships:
+                if rel.type in Relationship.PAIRED and str(rel.commit) not in filtered_in:
+                    filtered_in.remove(ref)
+                    break
```
What is the purpose of the `break` here and `:82`?

JonWBedard replied:
No reason to keep looking through the relationships of a commit if we've already determined the commit should be excluded.

alancoon commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
```diff
@@ -51,17 +52,36 @@ def parser(cls, parser, loggers=None):
+            if not relationships:
+                continue
+            for rel in relationships:
+                if rel.type in Relationship.PAIRED and str(rel.commit) not in filtered_in:
```
Can you explain `:71` thru `:82`?

My understanding: For every commit in `filtered_in`, get its relationships; for every relationship, check to see if any reverts or followups are not in `filtered_in`, if so, then remove the commit from `filtered_in` and break?

alancoon replied:
I don't understand why we'd want to disqualify a commit because we found a follow up?

JonWBedard replied:
Your explanation is correct, the "why" is important, though. By default, every commit on the branch is in `filtered_in` unless it's a cherry-pick, so, if a commit has a relationship that isn't in `filtered_in`, one of the following must be true:
- That commit has a relationship with a cherry-picked commit, and shouldn't be slated for merge-back
- That commit has a relationship with a commit on `main`, and shouldn't be slated for merge-back.

Put another way, the only commits we want to include are commits which don't have any relationships or have relationships with commits we're already including.

alancoon replied:
What about versioning commits? Are we filtering for them here?

alancoon replied:
Does this account for the cherry-picks that will occur to BPP branches?

JonWBedard replied:
We do filter versioning commits, yes, line 65.

I haven't totally taken into consideration BPP branches yet. BPP branches are only relevant when they're the branches we're scanning, which means we have to consider the branch we're branching from much the way the branch has to consider `main`. That being said, a BPP cherry-pick is already handled, because a BPP cherry-pick will refer to a branch commit which will refer to a `main` commit. The `main` commit shouldn't be in the branch story, therefor the whole chain of commits won't be slotted for cherry-pick.

alancoon commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/trace.py:
```diff
@@ -0,0 +1,276 @@
+        for type, regexes in {
+            cls.ORIGINAL: [CHERRY_PICK_RE],
+            cls.REVERTS: [REVERT_RE],
+            cls.FOLLOW_UP: FOLLOW_UP_FIXES_RE,
```
Why does `FOLLOW_UP_FIXES_RE` not match the `list` pattern of `CHERRY_PICK_RE` and `REVERT_RE`?

alancoon replied:
I see it's defined as a list at `:35`

alancoon commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/trace.py:
```diff
@@ -0,0 +1,276 @@
+        if commit.hash:
+            self.by_ref[commit.hash[:Commit.HASH_LABEL_SIZE]] = commit
+        if commit.revision:
+            self.by_ref['r{}'.format(commit.revision)] = commit
```
We're still writing tooling to deal with Trac-style revisions?

JonWBedard replied:
Not aggressively, but it's conceivable this tool (`git-webkit trace`, not `git-webkit pickable`) be use to deal with commits pre-transition.

dewei-zhu commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/trace.py:
```diff
@@ -0,0 +1,276 @@
+
+    def add(self, commit):
+        if str(commit) in self.by_ref:
+            return True
```
The return value seems a bit counter-intuitive.

alancoon replied:
It seems more natural to me if a successful add returned `True` and a no-op returned `False`.

dewei-zhu commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/trace.py:
```diff
@@ -0,0 +1,276 @@
+            cls.REVERTED_BY: cls.REVERTS,
+            cls.FOLLOW_UP: cls.FOLLOW_UP_BY,
+            cls.FOLLOW_UP_BY: cls.FOLLOW_UP,
+        }.get(type, type)
```
Any reason we want to return `type` for unknown `type` instead of error out?

JonWBedard replied:
My thinking is that if we don't define a pair, then the reverse of a relationship is itself, but probably `None` is more accurate.

dewei-zhu reviewed (approved):
r=me

JonWBedard committed:
[git-webkit] Add trace command
https://bugs.webkit.org/show_bug.cgi?id=244859
<rdar://97455957>

Reviewed by Dewei Zhu.

* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/__init__.py: Add Trace command.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
(Pickable.pickable): Create a CommitsStory object to track more commit relationships.
(Pickable.main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/trace.py: Added.
(Relationship.reversed): Reverse the relationship type (ie, the reverse of "reverted" is "reverted by").
(Relationship.parse): Parse a commit message and return relationships.
(Relationship.__init__):
(Relationship.__repr__):
(CommitsStory.issues_for): Parse a commit message for embedded issue references.
(CommitsStory.__init__):
(CommitsStory.add): Add a commit to the story.
(Trace.parser):
(Trace.relationships): Find all relationships for the specified commit.
(Trace.summary): Sumarize a commit.
(Trace.main): Find all relationships for the specified commit and print them.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/trace_unittest.py: Added.
(TestRelationship):
(TestCommitsStory):
(TestTrace):

Canonical link: https://commits.webkit.org/254997@main
